beacon-chain/operations/blstoexec/pool.go

```diff
@@ -0,0 +1,130 @@
+package blstoexec
+
+import (
+	"math"
+	"sync"
+
+	"github.com/prysmaticlabs/prysm/v3/config/params"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+)
+
+type listNode struct {
+	value *ethpb.SignedBLSToExecutionChange
+	prev  *listNode
+	next  *listNode
+}
+
+type changesList struct {
+	first *listNode
+	last  *listNode
+	len   int
+}
+
+// PoolManager maintains pending and seen BLS-to-execution-change objects.
+// This pool is used by proposers to insert BLS-to-execution-change objects into new blocks.
+type PoolManager interface {
+	PendingBLSToExecChanges(returnAll bool) []*ethpb.SignedBLSToExecutionChange
+	InsertBLSToExecChange(change *ethpb.SignedBLSToExecutionChange)
+	MarkIncluded(change *ethpb.SignedBLSToExecutionChange)
+}
+
+// Pool is a concrete implementation of PoolManager.
+type Pool struct {
+	lock    sync.RWMutex
+	pending changesList
+	m       map[types.ValidatorIndex]*listNode
+}
+
+// NewPool returns an initialized pool.
+func NewPool() *Pool {
+	return &Pool{
+		pending: changesList{},
+		m:       make(map[types.ValidatorIndex]*listNode),
+	}
+}
+
+// PendingBLSToExecChanges returns objects that are ready for inclusion at the given slot.
+// Without returnAll, this method will not return more than the block enforced MaxBlsToExecutionChanges.
+func (p *Pool) PendingBLSToExecChanges(returnAll bool) []*ethpb.SignedBLSToExecutionChange {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if returnAll {
+		result := make([]*ethpb.SignedBLSToExecutionChange, p.pending.len)
+		node := p.pending.first
+		for i := 0; node != nil; i++ {
+			result[i] = node.value
+			node = node.next
+		}
+		return result
+	}
+
+	length := int(math.Min(float64(params.BeaconConfig().MaxBlsToExecutionChanges), float64(p.pending.len)))
+	result := make([]*ethpb.SignedBLSToExecutionChange, length)
+	node := p.pending.first
+	for i := 0; i < length; i++ {
```

Comment:
This works, but somehow I think it's safer to make an empty slice and append on a for loop from first until node=last. Either way we need to have a safety check that the linked list wasn't broken or that there's no nil node in the middle.

Reply:
Looping until `node == last` won't work with a nil node in the middle. I will add a `node != nil` condition.